addons/payment_xendit/controllers/main.py

```diff
@@ -58,3 +58,12 @@ def _verify_notification_token(self, received_token, tx_sudo):
         if not consteq(tx_sudo.provider_id.xendit_webhook_token, received_token):
             _logger.warning("Received notification with invalid callback token %r.", received_token)
             raise Forbidden()
+
+    @http.route('/payment/xendit/payment', type='json', auth='public')
+    def xendit_payment(self, reference, partner_id, token_id):
+        """ Take newly created token, do a charge request and handle the response
+
+        :param str reference: The reference of the transation
+        """
+        tx_sudo = request.env['payment.transaction'].sudo().search([('reference', '=', reference)])
+        tx_sudo._xendit_create_charge(token_id)
```

Comment:
Shouldn't you rather create the token in Odoo and set Xendit's token_id as `provider_ref`, then call `_send_payment_request`?

Reply:
this is true. The reason I do this is, when I was developing it, the token created turns out to be invalid (because it was not multiple-use), which in this case results to unsuccessful payment, however the token is still created somehow.
So I'm doing this for safety reason to make sure that payment is successful first before storing the token in Odoo. Let me know what you think and if we should still proceed with changing the logic here

Reply:
that sounds like a good idea, will do that thanks!

Reply:
Hello @AntoineVDV for this one, I tried implementing it and have some concerns with the approach mentioned:

* we no longer have this _processPayment() function so I think it's not possible to model it after this SEPA one. 
* The possible option that I see is overriding the `_initiatePaymentFlow()`, create token, add it to paymentContext then update the flow to token. However I think with this approach, this means that we're forcing it to always save the payment method even if they don't tick that option when entering card details. 
* Payment processing cannot proceed with an archived token, so I don't think we can make it archived first.
* If we override rom the `_processDirectFlow`, the transactions is already created, not sure if it's a good idea to attach the token_id to transaction from this point on?

For these points, I believe it's best to keep the current implementation as is. Let me know if you think there's something better to do here. I checked all other payment methods now including SEPA and I don't think this way is currently implemented anymore...

Reply:
> - Payment processing cannot proceed with an archived token, so I don't think we can make it archived first.

That was my main concern and reason not to go this road.

Let's keep the current solution then ;)

> - we no longer have this _processPayment() function so I think it's not possible to model it after this SEPA one.
> - The possible option that I see is overriding the _initiatePaymentFlow(), create token, add it to paymentContext then update the flow to token. However I think with this approach, this means that we're forcing it to always save the payment method even if they don't tick that option when entering card details.
> - If we override rom the _processDirectFlow, the transaction is already created, not sure if it's a good idea to attach the token_id to transaction from this point on? If we do the _initiatePaymentFlow , this means we will create a new transaction again.

We have equivalent methods in 17.0 that could have met our requirements but no need to bother anymore if we create the token after payment.

Reply:
Noted with thanks! will resolve the convo in that case